### PR TITLE
feat(distributed): N7 host_orch + L3 notify/wait + allreduce end-to-end

### PR DESCRIPTION
**Pull request by YunjiQin**
## Summary

Brings up the N7 distributed pipeline end-to-end on NPU, taking the L3 notify/wait + allreduce examples from skipped placeholders to passing tests in ~7s on 2 ranks.

### New host-side codegen
- **host_orch.py emitter for N7** — `DistributedTensor` formals lower through `ContinuousTensor.make` + per-arg `add_scalar(contexts[r].device_ctx)` + `submit_next_level(worker=r)`; `pld.world_size()` lowers to `len(contexts)`.
- **`DistributedOpRegistry`** — per-op codegen hook (`REGISTER_DISTRIBUTED_OP`) mirroring `OrchestrationOpRegistry` / PTO `Backend::RegisterOp`. Handlers land for `tensor.slice`, `pld.tensor.alloc_window_buffer`, `pld.tensor.window`.
- **Per-op `VisitExpr_` overrides** for the 24 `BinaryExpr` subclasses, `UnaryExpr`, and `Cast` in the host_orch layer — base `IRVisitor` recurses operands without setting `current_expr_value_`, so `(r + 1) % t` previously collapsed to just `t`. New `distributed_scalar_expr_codegen.cpp` mirrors `pto_scalar_expr_codegen.cpp` semantics.
- **Scalar args wired through `EmitCallToWorker`** — the chip-dispatch arg loop previously skipped every `ScalarType` argument, silently dropping `peer` / `tag` from `self.chip_orch(..., peer, tag, device=r)`.

### N7 PTO codegen
- `pld.system.{get_comm_ctx, rank, nranks}` lower via a shared `EmitLoadRankPair` (i64 load of the `(rankId, rankNum)` slot, trunc / shrui+trunc to extract each half). IR rank/nranks return type INT32 → UINT32 to match runtime `CommContext::{rankId,rankNum}`.
- `pld.system.wait` emits the local view via `GetTensorViewTypeString` instead of materialising the concrete IR shape inline, keeping SSA value types consistent across uses.

### IR / pass fixes
- **SSA conversion** now walks `Call.attrs` values, not just `Call.args` — `attrs["device"] = ExprPtr` previously survived SSA versioning, leaving a dead `Var` that `CollectCommGroups` couldn't match. Promotes `\"device\"` to a `kAttrDevice` constant alongside the other attr-key constants.
- **`CollectCommGroups`** follows one level of SSA `AssignStmt` indirection when resolving the `ForStmt` loop bound (handles the `t = pld.system.world_size(); for r in pl.range(t):` form that `NormalizeStmtStructure` / CSE produces).
- **`tensor.read` / `tensor.write`** deducers + InCore property verifier + PTO codegen now accept `DistributedTensorType` via `AsTensorTypeLike` — scalar reads from a window-bound signal cell ride the same lowered path as a GM read, avoiding the 32-byte tile-alignment requirement.
- **`FunctionPassImpl::operator()`** preserves `program->comm_groups_` through the 4-arg `Program` ctor (previously every `FunctionPass` after `CollectCommGroups` silently wiped the freshly-collected groups).

### Runtime / linkage / defaults
- PTOAS emits `static` linkage for `CommRemoteOffset_i32` (`private` MLIR visibility on `func.func @CommRemoteOffset_<dtype>`) — was causing AIV mis-lowering and silent garbage in cross-rank pointer arithmetic.
- `DistributedConfig` defaults: `aicpu_thread_num` → 4 (tensormap+ringbuffer needs 3 schedulers + 1 dispatcher), `block_dim` → `None` (let L2 simpler pick its own).

### Tests
- `tests/st/distributed/test_l3_notify_wait.py` — restored to the design-doc `pl.range(pld.world_size())` loop form, no longer `pytest.mark.skip`.
- `tests/st/distributed/test_l3_allreduce.py` — new 2-rank port of `runtime/examples/workers/l3/allreduce_distributed`, exercising the full 4-phase pattern (stage-in, barrier, compute with `remote_load`, stage-out). Supersedes the standalone `remote_load` placeholder, which is removed.
- Two new UTs in `test_convert_to_ssa_pass.py` (`TestCallAttrSubstitution`) lock in the SSA `device=` attr substitution; two new UTs in `test_collect_comm_groups.py` cover the SSA-temp loop bound path.

## Testing
- [x] L3 notify/wait passes end-to-end on NPU 1,2 in 6.7s
- [x] L3 allreduce passes end-to-end on NPU 1,2 in ~7s
- [x] New SSA / `CollectCommGroups` UTs pass

### CODE REVIEW

File: tests/st/distributed/test_l3_allreduce.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            return self.reduce_step(inp, out, data, signal, peer)
+
+        @pl.function(level=pl.Level.HOST, role=pl.Role.Orchestrator)
+        def host_orch(
+            self,
+            inputs: pl.Tensor[[2, 1, SIZE], pl.FP32],
+            outputs: pl.Out[pl.Tensor[[2, 1, SIZE], pl.FP32]],
+        ) -> pl.Tensor[[2, 1, SIZE], pl.FP32]:
+            data_buf = pld.alloc_window_buffer(SIZE * 4)  # 1×SIZE × FP32 (4 bytes)
+            signal_buf = pld.alloc_window_buffer(4)  # 1×1 × INT32
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply tests/st/distributed/test_l3_allreduce.py]: Fixed in 72b6ca0d — replaced `×` with `x` in both comments.

File: src/ir/transforms/collect_comm_groups_pass.cpp [hunk 2]
```diff
@@ -140,11 +147,32 @@ class AllocAndWindowCollector : public IRVisitor {
...
+  ExprPtr cur = stop;
+  // Bounded chain walk; ``pld.system.world_size`` only ever produces one Call,
+  // so a depth of 4 is generous and guards against pathological cycles.
+  for (int depth = 0; depth < 4; ++depth) {
+    auto v = As<Var>(cur);
+    if (!v) return cur;
+    auto it = var_defs.find(v.get());
+    if (it == var_defs.end() || !it->second) return cur;
+    cur = it->second;
+  }
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/ir/transforms/collect_comm_groups_pass.cpp]: Fixed in 72b6ca0d — replaced the depth=4 bound with a cycle-safe `while` loop guarded by a visited-set of `Var*`s.

File: src/codegen/distributed/distributed_codegen.cpp [hunk 3]
```diff
@@ -464,6 +469,28 @@ void DistributedCodegen::VisitExpr_(const ir::CallPtr& op) {
...
+  // N7: ``pld.system.world_size()`` lowers to ``len(contexts)`` in host_orch
+  // Python. ``contexts`` is bound at every emitted orchestrator's signature
+  // (see EmitFunction / EmitEntryFunction) — for comm-less programs it's
+  // an empty list, which makes this a no-op fallback. Multi-group support
+  // would need a per-group ``len(contexts[group_idx])`` form; for the
+  // single-group N7 demo, the unindexed form is correct.
+  if (op->op_->name_ == "pld.system.world_size") {
+    current_expr_value_ = "len(contexts)";
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/codegen/distributed/distributed_codegen.cpp:584]: Acknowledged — the `contexts` invariant is already documented in the surrounding code comment; `pld.system.world_size` is only valid inside the host_orch entry where `contexts` is bound. Will tighten further if/when helper-scope lowering becomes a real use case.

File: src/ir/transforms/collect_comm_groups_pass.cpp [hunk 4]
```diff
@@ -156,18 +184,21 @@ DeviceDescriptor ResolveDeviceDescriptor(const ExprPtr& device, const std::vecto
...
         }
-        if (auto stop_ci = As<ConstInt>(fs->stop_)) {
+        if (auto stop_ci = As<ConstInt>(stop)) {
           int64_t start = 0;
-          if (auto start_ci = As<ConstInt>(fs->start_)) start = start_ci->value_;
+          if (auto start_ci = As<ConstInt>(UnwrapStopExpr(fs->start_, var_defs))) start = start_ci->value_;
           int64_t step = 1;
-          if (auto step_ci = As<ConstInt>(fs->step_)) step = step_ci->value_;
+          if (auto step_ci = As<ConstInt>(UnwrapStopExpr(fs->step_, var_defs))) step = step_ci->value_;
           CHECK(step == 1) << "CollectCommGroups: device=r over a non-unit-step loop is not supported "
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/ir/transforms/collect_comm_groups_pass.cpp:203]: Fixed in 72b6ca0d — `start` and `step` now require `ConstInt` via `CHECK`, symmetric with the existing `stop` validation.

File: src/codegen/distributed/distributed_ops_codegen.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      // Drop this axis via scalar indexing — torch reduces the rank by 1.
+      indices << offset_i;
+    } else {
+      const std::string shape_i = codegen.GetExprAsCode(shape_tuple->elements_[i]);
+      // Constant-fold ``0 + shape_i`` for readability when offset is 0.
+      auto offset_const = As<ConstInt>(offset_tuple->elements_[i]);
+      if (offset_const && offset_const->value_ == 0) {
+        indices << "0:" << shape_i;
+      } else {
+        indices << offset_i << ":" << offset_i << " + " << shape_i;
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/codegen/distributed/distributed_ops_codegen.cpp:130]: Skipping — `offset_i` / `shape_i` come from `MakeTuple` elements whose non-emptiness is enforced at the IR layer; an empty render at this point would indicate an IR construction bug rather than a codegen edge case.

File: tests/ut/codegen/test_distributed_codegen.py [hunk 6]
```diff
@@ -44,7 +44,8 @@ def host_orch(self, x: pl.Tensor[[64], pl.FP32]) -> pl.Tensor[[64], pl.FP32]:
         code = cg.generate(program)
 
         # Verify imports
-        assert "from simpler.task_interface import TaskArgs, TensorArgType" in code
+        assert "from simpler.task_interface import " in code
+        assert "TaskArgs" in code and "TensorArgType" in code
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply tests/ut/codegen/test_distributed_codegen.py:48]: Skipping — the current substring checks are intentionally loose to avoid regex-tracking the exact import form (which has changed twice already). The downstream behavior is exercised by the broader codegen tests; tightening here would add maintenance cost without catching a realistic bug class.

File: src/backend/common/pto_ops_common.cpp [hunk 7]
```diff
@@ -2804,6 +2812,65 @@ void RegisterPTOOps(Backend& backend, const std::unordered_set<std::string>& exc
...
+    namespace cl = codegen::distributed::comm_layout;
+    static_assert(cl::kRankNumOffset == cl::kRankIdOffset + 4,
+                  "pld.system.nranks codegen assumes rankNum sits in the high 32 bits of rankId's i64 slot");
+    std::string ctx_ssa = cg.GetExprAsCode(op->args_[0]);
+    std::string rk_pair = EmitLoadRankPair(cg, ctx_ssa);
+    std::string c32 = cg.GetOrEmitConstant(static_cast<int64_t>(32), DataType::INT64);
+    std::string rn_i64 = cg.NewTemp();
+    cg.Emit(rn_i64 + " = arith.shrui " + rk_pair + ", " + c32 + " : i64");
+    std::string rn = cg.NewTemp();
+    cg.Emit(rn + " = arith.trunci " + rn_i64 + " : i64 to i32");
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/backend/common/pto_ops_common.cpp:3070]: Skipping — the current `arith.trunci ... : i64 to i32` form works end-to-end on the target backend, and consumers don't yet read the IR-level UINT32 dtype. Logged as a low-severity latent issue in `KNOWN_ISSUES.md` for follow-up.

File: src/codegen/distributed/distributed_codegen.cpp [hunk 8]
```diff
@@ -774,6 +859,55 @@ std::string DistributedCodegen::DataTypeToPythonDType(const DataType& dtype) {
...
+  return "FLOAT32";
+}
+
+std::string DistributedCodegen::ResolveRankExpr(const ir::CallPtr& call) const {
+  if (!call->HasAttr(ir::kAttrDevice)) return "";
+  auto dev = call->GetAttr<ir::ExprPtr>(ir::kAttrDevice, nullptr);
+  INTERNAL_CHECK_SPAN(dev != nullptr, call->span_) << "device= attr must hold a non-null ExprPtr";
+  if (auto ci = std::dynamic_pointer_cast<const ir::ConstInt>(dev)) {
+    return std::to_string(ci->value_);
+  }
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/codegen/distributed/distributed_codegen.cpp:1003]: Fixed in 72b6ca0d — added `CHECK(ci->value_ >= 0)` in the ConstInt branch of `ResolveRankExpr`.

File: src/codegen/distributed/distributed_ops_codegen.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      << "tensor.slice host_orch codegen expects 3-5 args (input, shape, offset[, valid_shape[, "
+         "drop_dims]]), "
+         "got "
+      << op->args_.size();
+
+  const std::string input_name = codegen.GetExprAsCode(op->args_[0]);
+  CHECK(!input_name.empty()) << "tensor.slice input must resolve to a non-empty Python name";
+
+  const std::string lhs = codegen.GetCurrentResultTarget();
+  CHECK(!lhs.empty()) << "tensor.slice in host_orch must have an assignment target";
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/codegen/distributed/distributed_ops_codegen.cpp:96]: Keeping CHECK — FlattenCallExpr is contracted to hoist tensor.slice into AssignStmt position before host_orch codegen runs, so this is the right guard for that invariant.

File: src/codegen/pto/pto_codegen.cpp [hunk 10]
```diff
@@ -1416,8 +1414,15 @@ void PTOCodegen::VisitStmt_(const AssignStmtPtr& op) {
...
-  if (As<ScalarType>(op->var_->GetType()) && !fs_.current_expr_value.empty()) {
+  // Register scalar/index/CommCtx result so subsequent expressions can look up
+  // this variable. N7: CommCtxType is a singleton marker; the bound SSA is the
+  // matching ``!pto.ptr<i64>`` ctx ptr from the func.func trailing-ctx segment
+  // (no MLIR is emitted for ``pld.system.get_comm_ctx`` — its lambda just sets
+  // ``current_expr_value`` to the ctx SSA). Treating it like a scalar here lets
+  // downstream ``pld.system.rank(ctx)`` / ``pld.system.nranks(ctx)`` codegen
+  // resolve ``ctx`` via the standard ``GetExprAsCode(call->args_[0])`` path.
+  const auto& var_type = op->var_->GetType();
+  if ((As<ScalarType>(var_type) || As<CommCtxType>(var_type)) && !fs_.current_expr_value.empty()) {
```
[the remark this answers is not in the document]
Comment (YunjiQin) [reply src/codegen/pto/pto_codegen.cpp:1438]: Fixed in 72b6ca0d, but in the op handlers rather than `VisitStmt_(AssignStmt)`: `get_comm_ctx` now calls `RegisterVarToMlir(GetCurrentResultVar(), ctx_ssa)` to override the Branch-A pre-allocated temp, and `rank` / `nranks` now emit into `GetCurrentResultTarget()` (matching the standard pattern used by every other backend handler in this file).

**Review by lyfne123 — APPROVED** [review]
